### ISSUE TITLE
ES6 Class for protocol.js Fronts

### ISSUE THREAD

**Issue by ochameau** [devtools-html/rfcs#56]
Switching protocol.js to use es6 class is possible! But as Firefox doesn’t (yet) support fields, there is two features that we can’t easily migrate as-is. But I’m convinced this is a good opportunity to simplify these two quite cryptic features, and, at the same time switch to classes and benefit from builtin inheritance.

Fields in es6 class are functions defined like this:
```js
class MyFront {
  myMethod = function () {  // This doesn’t work in Firefox
  }
}
```
Instead of the regular class functions:
```js
class MyFront {
  myMethod() {

  }
}
```

This prevents migrating the usage of prococol.custom and protocol.preEvent methods:
```js
  waitForHighlighterEvent: protocol.custom(function(event) {
    return this._waitForHighlighterEvent(event, this.toolbox.highlighter.actorID);
  }, {
    impl: "_waitForHighlighterEvent",
  }),
```
```js
  valueChange: preEvent("value-change", function(value) {
     this._form.value = value;
  }),
```
Here is proposals on how to migrate these two features to es6 classes.

### Overloading RDP requests, a.k.a “custom”
Sometimes, you have to compute RDP request arguments automatically by the front, or have to interpret the RDP request’s response. To do that, you are overloading the RDP request, but you also have to wrap the function via `custom` function.
With the new setup, you would no longer rely have to go through `custom` function and simply overload the RDP request via es6 class and call the original method via `super`.
```diff
-  waitForHighlighterEvent: protocol.custom(function(event) {
-    return this._waitForHighlighterEvent(event, this.toolbox.highlighter.actorID);
-  }, {
-    impl: "_waitForHighlighterEvent",
-  }),
```
```diff
+  waitForHighlighterEvent(event) {
+    return super.waitForHighlighterEvent(event, this.toolbox.highlighter.actorID);
+  }
```
### Pre-processing of events, a.k.a “preEvent”
Sometimes, you have to process the events emitted by the actor and be sure that this processing is done before any other event listener is called. To do that, you have to implement a method on the Front with an arbitrary name and wrap it via `preEvent` method.
With the new setup, you would no longer rely have to use the `preEvent` function and simply register the processing function via `registerPreEventListener`:
```js
   constructor(client, form) {
     super(client, form);
+
+    this.registerPreEventListener("name-change", this.nameChange.bind(this));
   }
-  valueChange: preEvent("value-change", function(value) {
+  valueChange(value) {
     this._form.value = value;
-  }),
+  )
```

- One alternative to registerPreEventListener would be to tweak or subclass EventEmitter in order to support an options object and do this:
```js
+    this.on("name-change", this.nameChange.bind(this), { preEvent: true });
```
`preEvent` would mean that the listener has to be called first and wait for its completion before calling any other non-preEvent listener. Obviously, `preEvent` isn’t such a good name and we could probably find a better naming for it.
As suggested by Nicolas, instead of augmenting `on`, we could introduce a new function on `EventEmitter` called `before` and use it like this:
```js
+    this.before("name-change", this.nameChange.bind(this));
```
I originally suggested augmenting `on` in order to mimic what we are doing in `addEventListener` where we introduce an options argument allowing to control how we do listen for the given event. Like `addEventListener("click", listener, { once: true, capture: true })`.

- Use the specification to define a method name on the front that is a pre-event handler. (Yura’s proposal)
In the specification:
```js
  events: {
    "name-change": {
      type: "name-change",
      frontPreHandler: “nameChange”
    },
```
In the front class
```js
  valueChange(value) {
    this._form.value = value;
  }),
```
One issue with this approach is that we start having front specifics in the specification, whereas so far, all specification fields are about RDP packet layout and how it is mapped to both Actors and Fronts.

### Migration from initialize to es6 constructors and use es6 extends instead of SDK’s `extend` 
Today, `FrontClassWithSpec` uses SDK’s `extend` helper in order to compose from multiple classes: the `Front` base class, the class generated out of the specification and the specific class (i.e. the second argument passed to FrontClassWithSpec).
But, instead, we could use es6 class inheritance and make the specific class like `InspectorFront` inherit from a mixin generated by FrontClassWithSpec.
Here is how the typical front class definition would look like:
```js
-var InspectorFront = FrontClassWithSpec(inspectorSpec, {
-  initialize: async function(client, form) {
-    Front.prototype.initialize.call(this, client, form);
-    ...
-  },
-  _getWalker: async function() {
-     ...
-  },
```
```js
+class InspectorFront extends FrontClassWithSpec(inspectorSpec) {
+  async constructor(client, form) {
+    super(client, form);
+    ...
+ }
+  async _getWalker() {
+     ...
+  }
```
### Keep support of async front setup
Some fronts like Inspector and Performance require some asynchronous actions during their setup. The inspector for example, has to retrieve the Walker and Highlighter front by doing a RDP request. In order to do that, we recently introduced async initialize method.
```js
var InspectorFront = FrontClassWithSpec(inspectorSpec, {
  initialize: async function(client, tabForm) {
    ...
    // async initialization
    await Promise.all([
        this._getWalker(),
        this._getHighlighter(),
    ]);
}
```
So that, when we do target.getFront(“inspector”) we retrieve a Promise that is resolved only once InspectorFront.initialize method is resolved.
Now that initialize is the es6 class constructor, we can no longer use async functions. es6 classes do not support having async constructor function. But they still support returning a Promise out of it.
So that we could do that:
```js
class InspectorFront extends FrontClassWithSpec(inspectorSpec) {
  constructor(client, form) {
    super(client, form);
    ...
    return this.initialize();
  }
  initialize: async function() {
    // async initialization
    await Promise.all([
        this._getWalker(),
        this._getHighlighter(),
    ]);
   return this; // This is important, otherwise the promise resolve to a null value,
                // and there is no other way to retrieve the front instance.
  }
}
```
Note that this is an optional feature, only front which require an asynchronous setup would have to do that.

**Comment by codehag** [comment]
Thanks for the write up Alex, this is really fantastic work!

From my perspective, there are three topics to discuss here:
1) protocol method overloads (custom)
2) event overloads (preEvent)
3) async initialization

1 and 3 are simpler so I will start with those

**Topic 1** 

I think is very nice, and can be considered solved. The only issue that I see with this approach is that it is no as obvious that something is overloading the RDP. You have to look more carefully to see that something is custom. But this is really minor, I quite like where this is going!

**Topic 3** 

is something that I was expecting, and I like the approach you took. We continue to operate on the assumption that all fronts are async, and that we need to await them. I am only concerned about how this will impact marshalling, as we haven't addressed that yet. This will only work with fronts retrieved via getFront. However, what I like about this is we can easily split things up if need be. I also like that the constructor will not technically be async, just it's return value.

**Topic 2** 
preEvents i think is the stickiest thing to discuss. We have 3~4 options

1) use this.registerPreEventListener (it looks like it would be something that we need to write, and would exist on the front?)

I am guessing that this would be defined on the front prototype, and there is something nice about it. I think we could even apply the naming that nicolas suggested, however we would be duplicating the functionality of the eventemitter in part. 

2) modify the Event emitter
2.a) to take an options object
2.b) to have a before method

Of those two option i prefer the `this.before` naming. This way the `on` function is kept simple. The issue with this, is that before is not necessary anywhere else. If we can imagine a world in which it is useful for other modules that use the EventEmitter, then this might be a good idea.

 3) modify the spec to indicate a preEvent

I don't like this option as much, as I see the spec as defining the relationship between the client and the server, not their individual specifics. I am afraid that if we start adding front specifics, we will then start adding server specifics, and concerns will be tangled.

Out of these three options **I am leaning towards option 2.b**, as the syntax is really intuitive, and it feels like something that belongs to the event emitter. However I can be convinced otherwise. For example if it makes the event emitter too complex for other applications, or provides a crutch that can lead to messy code patterns, though I can't think of any at the moment.

**Comment by jasonLaster** [comment]
This proposal makes sense to me. As an added benefit the new version feels simpler.

**Comment by ochameau** [comment]
I'm open to more feedback about "pre events", but I would like to suggest as a first iteration to introduce `before` and only implement it on `Front` class. Implementing this in `EventEmitter` is slightly more complex (would make `emit` be async) and it would require some more feedback to confirm it is useful enough to be part of this very common class.

**Comment by ochameau** [comment]
@juliandescottes in [[nosdk] use constructor instead of initialize in actors and fronts that inherit Actor/Front classes](https://bugzilla.mozilla.org/show_bug.cgi?id=1393615)
> I gave this a shot today and switching fronts and actor to ES6 classes will be challenging for another reason. We are using protocol.custom() and protocol.preEvent() to decorate some methods of Front prototypes. 
>
> I think we will have to keep using the old extend() approach for the time being.

It looks like you approve the switch to es6 class and use the constructor instead of initialize.
I'm now looking forward your feedback as you went into custom and preEvent :)

I opened [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1510949) to track the refactoring of fronts and [this other one](https://bugzilla.mozilla.org/show_bug.cgi?id=1510951) to refactor the actors. As it should be easier to refactor actors (but that isn't strictly necessary for fission).

**Comment by juliandescottes** [comment]
Thanks for the proposal! Completely in favor of moving to es6 classes and to simplify this. As you've seen I gave up during the SDK removal because this was non-trivial but I think your propositions make sense. I will use the breakdown from @codehag to comment on the different parts.

**topic 1: protocol method overloads (custom)**

Not much to say, I agree that calling super's implementation seems like the way to go.

**topic 2: preEvent**

I would prefer adding a method to the Fronts, because I find the API risky in general.

Supporting "before" for events sounds similar to using "important" in CSS. It doesn't map to a real phase of the event (as `capture` does). Once two call sites start using `before` for the same event, then you no longer are guaranteed to be called "first". 

Also the fact that we need to wait for the `before` listeners to resolve before calling regular listeners also seems like a dangerous concept to introduce for any event. When you fire an event you can expect it to be processed by the listeners quickly, even though you should not rely on it. Here it's no longer true. Should each before listener be called in sequence, should we call them in parallel?  

The event-emitter logic is really small, and I don't think we need the whole set of feature to just implement `registerPreEventListener` on Fronts.

I am **in favor of option 1**.

**topic 3: async constructors**

There are many ways to address this, none perfect. Overall your suggestion is ok but I see two things that might be confusing here, so I'll share my feedback on them :) 

First, the difference between the `constructor` and `initialize`. If a Front has code in both constructor and initialize, how do you decide what should go in the constructor and what should go in initialize? Maybe it could be nice to completely forget about constructors, and only ask Fronts to implement initialize? Possibly enforced with an eslint rule.

Then the "return this" mandatory in the initialize. I think that's easy to forget about and also hard to understand. On the other hand I think that most of the code that will instantiate Fronts will be centralized in the framework? Do we expect to manually create fronts frequently? Between `const front = await new Front();` and `const front = new Front(); await front.initialize();` my preference goes to the second one unless this code is really duplicated in the whole codebase. If you'd like we could even have a static helper on the base Front class called `async createFront`. 

With both these points in mind, here's another take on this proposal: https://gist.github.com/juliandescottes/5424b841f9991068cda2aeaff678e12a . Most of this is probably down to personal preference anyway. If you like some of the ideas, feel free to adapt them. Otherwise as I said I'm overall OK with the suggestion.

**Comment by ochameau** [comment]
> The event-emitter logic is really small, and I don't think we need the whole set of feature to just implement registerPreEventListener on Fronts.
>
> I am in favor of option 1.

What about implementing `before` on Front class? It would be like `registerPreEventListener`, but named `before`.
Note that in my current implementation, I support registering only one listener. The previous code was used to support more than one, but as you said, it introduces hazardous call order. And we always register one such listener for each event, I can't see why we would register more than one.

> Between `const front = await new Front();` and `const front = new Front(); await front.initialize();` my preference goes to the second one unless this code is really duplicated in the whole codebase.

+1 I'll open a followup bug to polish the async fronts. I think we should be able to do:
```
const front = new Front();
await front.initialize();
```
from all callsites creating fronts.

**Comment by juliandescottes** [comment]
Looks like all the open questions have been resolved and I think the related work already landed in [Bug 1510949](https://bugzilla.mozilla.org/show_bug.cgi?id=1510949)
